test/distributed/fsdp/test_fsdp_hybrid_shard.py

```diff
@@ -313,11 +317,73 @@ def patched_collective(orig_collective, counter, *args, **kwargs):
             self.assertEqual(num_hsdp_flat_params, cntr[orig_ar])
             self.assertEqual(num_flat_params, cntr[orig_rs])
 
+    @skip_if_lt_x_gpu(4)
+    def test_fsdp_hybrid_shard_parity(self):
+        self.run_subtests(
+            {
+                "hsdp_sharding_strategy": [
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
+                ],
+                "use_orig_params": [False, True],
+            },
+            self._test_fsdp_hybrid_shard_parity,
+        )
+
+    def _test_fsdp_hybrid_shard_parity(
+        self, hsdp_sharding_strategy: ShardingStrategy, use_orig_params: bool
+    ):
+        fsdp_model = self._init_fsdp_model(use_orig_params)
+        global_pg = dist.distributed_c10d._get_default_group()
+        hsdp_pgs = _init_intra_and_inter_node_groups(global_pg, 2)
+        hsdp_model = self._init_hsdp_model(
+            hsdp_sharding_strategy,
+            ShardingStrategyMode.ALL_HYBRID_SHARD,
+            use_orig_params,
+            hsdp_process_groups=hsdp_pgs,
+        )
+        assert (
+            hsdp_model._inter_node_pg.size() > 1
+        ), "HSDP model initialized without replication"
+        fsdp_optim = torch.optim.Adam(fsdp_model.parameters(), lr=1e-2)
+        hsdp_optim = torch.optim.Adam(hsdp_model.parameters(), lr=1e-2)
+        for _ in range(5):
+            inp = fsdp_model.module.get_input(torch.device("cuda"))
```

Comment:
you might have to seed differently across ranks to ensure that inputs are diff across ranks for this test to be nontrivial?

Reply:
Fixed!